crates/wasmtime/src/store.rs

```diff
@@ -123,12 +126,52 @@ impl Store {
                 memory_count: Default::default(),
                 table_count: Default::default(),
                 fuel_adj: Cell::new(0),
+                context_values: RefCell::new(HashMap::new()),
             }),
         }
     }
 
-    pub(crate) fn from_inner(inner: Rc<StoreInner>) -> Store {
-        Store { inner }
+    /// Gets a function from the [`Engine`] associated with this [`Store`].
+    ///
+    /// Returns `None` if the given function is not defined.
+    pub fn get_engine_func(&self, module: &str, name: &str) -> Option<Func> {
+        let func = self.inner.engine.get_func(module, name)?;
+
+        let mut funcs = self.inner.engine_funcs.borrow_mut();
+
+        let anyfunc = funcs
+            .entry(unsafe { func.instance.clone() })
+            .or_insert_with(|| {
+                let mut anyfunc =
+                    match func
+                        .instance
+                        .lookup_by_declaration(&wasm::EntityIndex::Function(
+                            wasm::FuncIndex::from_u32(0),
+                        )) {
+                        Export::Function(f) => unsafe { f.anyfunc.as_ref() }.clone(),
+                        _ => unreachable!(),
+                    };
+
+                // Register the function with this store's signature registry
+                anyfunc.type_index = self
+                    .inner
+                    .signatures
+                    .borrow_mut()
+                    .register(func.ty.as_wasm_func_type(), func.trampoline);
+
+                Box::new(anyfunc)
+            });
+
+        let instance = StoreInstanceHandle {
+            store: self.clone(),
+            handle: func.instance,
```

Comment:
This is one of the things I'm most worried about with this PR. I don't think there's any issue with this currently, only that one could be lurking by accident in the future.

My concern here is that the not-threadsafe `InstanceHandle` is being shared across threads. This is ok because we happen to not touch any of the non-threadsafe state, but all of `InstanceHandle` and consumers are currently written without thread-safety in mind. Do you think there's anything we can do to future-proof this? For example I mentioned one elsewhere in this PR where we store an invalid index so if it's ever used we hopefully abort quickly. I think ideally we'd have checks in `wasmtime_runtime::Instance` as well, but I'm not sure if they would be too onerous for other usages.

Basically my main worry here is that this function is introducing threading in a non-compiler-checked fashion. All authors of `runtime/src/instance.rs` need to be aware of this but that's difficult to enforce because this happens in such an isolated location.

Reply:
> My concern here is that the not-threadsafe `InstanceHandle` is being shared across threads. This is ok because we happen to not touch any of the non-threadsafe state, but all of `InstanceHandle` and consumers are currently written without thread-safety in mind. Do you think there's anything we can do to future-proof this?

Would it be feasible to split out a struct of `Sync` members from `InstanceHandle` that we can share, without sharing the whole thing across threads?

e.g.

```rust
struct InstanceHandleSync {
    // sync members in here...
}

struct InstanceHandle {
    sync: InstanceHandleSync,
    // non-sync members here...
}
```

Reply:
Unfortunately I don't think so since most JIT code goes from `*mut VMContext` straight to `*mut Instance` (the internals of the handle) which has all the non-`Sync` pieces. For example we just have a dynamic guarantee that the instance never tries to `elem.drop`, for example.

One of my bigger worries is leaking the incorrect `VMSharedSignatureIndex` which is configured in the `VMContext`. I think storing an out-of-bounds index will help weed things out, but otherwise I don't have any great ideas. That being said I don't believe that it's leaked today at all.

Reply:
I share your concerns over safety here.

I still think we're getting some mileage out of the compiler for safety as `Store` is `!Send+!Sync` and the store to use for an engine function is always retrieved from TLS (and thus should be the same thread that created the store).

Perhaps a check in `Caller::store()` that ensures the call is happening on the same thread that created the store, just to be safe?

> Would it be feasible to split out a struct of Sync members from InstanceHandle that we can share, without sharing the whole thing across threads?

I think splitting `wasmtime_runtime::Instance` like that shouldn't be necessary as the thread-affinity of an instance shouldn't be violated by these changes.

Reply:
My worry is moreso for the internals of Wasmtime itself rather than callers. You're right though in that we do have a number of consistency checks and internals (I suspect more than a few locations modified in this PR were found via assertion failures rather than knowing what to do off the top of your head). My biggest worry is hidden mutability in `InstanceHandle`, but I think it's not the end of the world for that to be something we remain vigilant for.

Perhaps a nice doc-block can go somewhere though explaining the threading involved here?

Reply:
I've added a comment where we have the unsafe impls for `HostFunc`.